File: packages/solutions/python-server/trainer.py
from sklearn.metrics import classification_report, confusion_matrix
import tensorflow as tf
from model import mobilenet_v3_large
from utils import load_data
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
import wandb

image_height = 98
image_width = 128
image_channel = 3
batch_size = 20
epochs = 100
patience = 10
num_classes = 7


def load():
    train_x, train_y = load_data(
        'packages/datasets/327labeled CK+', image_height, image_width, image_channel, num_classes)
    train_x, test_x, train_y, test_y = train_test_split(
        train_x, train_y, test_size=0.3, random_state=42)
    # test_x, test_y = load_data('packages/datasets/images/validation',
    #

    print('test', test_x)
    # image_height, image_width, image_channel, num_classes)
    return train_x, train_y, test_x, test_y


def train(train_x, train_y, test_x, test_y):
    model = mobilenet_v3_large(
        input_shape=(image_height, image_width, image_channel),
        num_classes=num_classes,
        include_top=True
    )

    model.summary()

    model.compile(
        optimizer=tf.keras.optimizers.Adam(0.0005),
        loss=tf.keras.losses.CategoricalCrossentropy(),
        metrics=['accuracy'],
    )

    history = model.fit(
        train_x,
        train_y,
        batch_size=batch_size,
        verbose=1,
        epochs=epochs,
        validation_data=(test_x, test_y),
        callbacks=[
            tf.keras.callbacks.EarlyStopping(
                monitor='val_loss', patience=patience, restore_best_weights=True
            ),
        ]
    )

    # 手动记录训练过程中的损失和准确率
    for epoch in range(len(history.history["loss"])):
        wandb.log({"train_loss": history.history["loss"][epoch],
                   "train_accuracy": history.history["accuracy"][epoch]})
        wandb.log({"val_loss": history.history["val_loss"][epoch],
                   "val_accuracy": history.history["val_accuracy"][epoch]})
    return model, history


def draw_loss(history):
    # 提取训练和验证损失的历史记录
    training_loss = history.history['loss']
    validation_loss = history.history['val_loss']
    # 获取训练的epoch数量
    epochs = range(1, len(training_loss) + 1)
    # 绘制训练和验证损失曲线
    plt.figure(figsize=(10, 6))
    plt.plot(epochs, training_loss, label='Training Loss')
    plt.plot(epochs, validation_loss, label='Validation Loss')
    plt.xlabel('Epochs')
    plt.ylabel('Loss')
    plt.legend()
    plt.title('Training and Validation Loss Over Epochs')
    plt.grid(True)
    plt.savefig('packages/solutions/python-server/losses/loss_{}.png'.format(
        history.history['val_accuracy'][-1]))

    plt.show()


def main():
    train_x, train_y, test_x, test_y = load()
    wandb.init(project="your_project_name",
               name="mobilenet_v3_large",
               config={
                   "image_height": image_height,
                    "image_width": image_width,
                    "image_channel": image_channel,
                    "batch_size": batch_size,
                    "epochs": epochs,
                    "patience": patience,
                    "num_classes": num_classes,
               })
    model, history = train(train_x, train_y, test_x, test_y)
    print(history)
    model.save('./packages/solutions/python-server/models/MobileNetV3_{}.h5'.format(
        history.history['val_accuracy'][-1]))
    draw_loss(history)
    wandb.finish()


if __name__ == '__main__':
    main()

File: packages/solutions/js-static-model/trainer.ts
import * as tf from "@tensorflow/tfjs-node";
import mobileNetV3Large from "./model";
import { EarlyStopping, Optimizer, Tensor3D } from "@tensorflow/tfjs-node";
import { loadData, trainTestSplit } from "./utils";
import wandb from "@wandb/sdk";

interface LabeledData {
  imagePath: string;
  label: number;
}

const imageHeight = 98,
  imageWidth = 128,
  imageChannel = 3,
  batchSize = 5,
  epochs = 2,
  patience = 10,
  numClasses = 7;

async function train() {
  wandb.init({
    project: "your_project_name",
    name: "mobilenet_v3_large_js",
    config: {
      imageHeight,
      imageWidth,
      imageChannel,
      batchSize,
      epochs,
      patience,
      numClasses,
    },
  });

  const [imagesPromises, labels] = loadData(
    "../../datasets/327labeled CK+",
    imageHeight,
    imageWidth,
    imageChannel,
    numClasses
  );
  const images = await Promise.all(imagesPromises);
  // 将图像Tensors堆叠成一个批次
  const trainX = tf.stack(images);
  const trainY = tf.oneHot(labels, numClasses);

  const model = mobileNetV3Large({
    inputShape: [imageHeight, imageWidth, imageChannel],
    alpha: 1.0,
    includeTop: true,
    numClasses,
  });
  model.summary();

  const optimizer = tf.train.adam(0.0005);
  model.compile({
    optimizer,
    loss: "categoricalCrossentropy",
    metrics: ["accuracy"],
  });

  const history = await model.fit(trainX, trainY, {
    batchSize,
    verbose: 1,
    epochs,
    validationSplit: 0.3,
    callbacks: [
      new EarlyStopping({
        monitor: "val_loss",
        patience,
      }),
    ],
  });

  for (const epoch in history.history["loss"]) {
    console.log(history, history.history);
    wandb.log({
      train_loss: history.history["loss"]?.[epoch] ?? 0,
      train_accuracy: history.history["accuracy"]?.[epoch] ?? 0,
    });
    wandb.log({
      val_loss: history.history["val_loss"]?.[epoch] ?? 0,
      val_accuracy: history.history["val_accuracy"]?.[epoch] ?? 0,
    });
  }

  console.log(history);
  await model.save(
    `file://./packages/solutions/js-static-model/models/model_${history.history?.["accuracy"] ?? "unknown"}`
  );
  wandb.finish();
}

train();
